Mask load/save: Support numpy and edf format and vertically flipped mask

t20100 opened the pull request:
This PR aims at making the mask interoperable with other tools:
- Adding support for load/save mask from/to npy and edf.
- Saving the mask not flipped with respect to original data (and also allowing to load it), by providing extra options in the dialog:
  - Mask (*.mask): Unchanged behavior
  - Vertically flipped mask (*.npy *.edf): ~~Reflecting image flip applied on load to mask load/save~~ Flip mask vertically on load/save

I also started looking at a "vertically flipped (*.mask)" option: There is a TIFF "Orientation" tag to store this information. `Pillow` is interpreting it and returns the flipped array.. so it does not really help. I have left this aside for now.

~~Regarding the vertical flip, it depends on the image file loader that was used: It is only applied when loaded with PIL, fabio or lambda loader (and not with spe, karabo and h5). So since the goal is to have a mask not flipped with respect to the original data, I store the flip information in the image model and use it to flip or not the mask for load/save from the `MaskController`. This made this PR bigger than first expected.~~

I haven't yet looked at the batch yet, and since it calls `load_mask`, it will need to be adapted as well (so I leave this PR as Draft).

t20100 commented on dioptas/model/ImgModel.py:
```diff
@@ -71,6 +71,7 @@ def __init__(self):
         self._img_data_background_subtracted = None
         self._img_data_absorption_corrected = None
         self._img_data_background_subtracted_absorption_corrected = None
+        self.img_flipud = False
```
Stored by loader to be retrieved later

t20100 commented on dioptas/model/MaskModel.py:
```diff
@@ -231,24 +232,44 @@ def set_mask(self, mask_data):
         self._mask_data = mask_data
 
-    def save_mask(self, filename):
+    def save_mask(self, filename, flipud: bool=False):
```
Adding a `flipud` argument so the mask model does not need access to the image model.

t20100 commented on dioptas/model/MaskModel.py:
```diff
@@ -231,24 +232,44 @@ def set_mask(self, mask_data):
+        if flipud:
+            im_array = np.flipud(im_array)
+
+        if filename.endswith('.npy'):
+            np.save(filename, im_array)
+        elif filename.endswith('.edf'):
+            fabio.edfimage.EdfImage(im_array).write(filename)
+        else:
```
I put new formats first, so it falls back to the previous implementation

t20100 commented on dioptas/model/MaskModel.py:
```diff
@@ -231,24 +232,44 @@ def set_mask(self, mask_data):
-            data = np.array(Image.open(filename))
-        except IOError:
-            data = np.loadtxt(filename)
+    def _load_mask(self, filename, flipud: bool):
```
Sharing loading implementation between `load_mask` and `add_mask`

t20100 commented on dioptas/tests/controller_tests/test_MaskController.py:
```diff
@@ -57,11 +61,14 @@ def test_loading_and_saving_mask_files(self):
-        QtWidgets.QFileDialog.getSaveFileName = MagicMock(return_value=filename)
-        click_button(self.mask_widget.save_mask_btn)
-        self.assertTrue(os.path.exists(filename))
+        for extension in self.MASK_EXTENSIONS:
```
I adapted existing to also test new mask extensions

t20100 commented on dioptas/tests/functional_tests/test_mask_save_and_load.py:
```diff
@@ -0,0 +1,156 @@
+from ..utility import click_button, unittest_data_path, QtTest
+from ...controller.MainController import MainController
+
+
```
I made this new test with pytest fixtures. If you prefer that it's use `QtTest`, I'll change this.

CPrescher commented:
Hi, this looks good to me. I have only small remarks:

- _load_mask can be outside of the mask class (not dependent on class) and be renamed to read_mask_file (load implies for me, that it is put into some variable inside of a class, read would imply it is returned as value)
- all functions with the flipud parameter should have documentation, since it is not immediately obvious what the parameter means
- for read_mask_file the flipud parameter should probably also be default
- small note regarding styling: my linter wants to have spaces around "=" in the function declaration - do you use different styles?
- pyfixtures are perfect, i did not know about them when I started in 2013, but will write most of the tests now with fixtures

Thank you!

CPrescher commented on dioptas/controller/MaskController.py:
```diff
@@ -368,24 +368,35 @@ def cosmic_btn_click(self):
+    def _mask_filters(self) -> str:
+        if self.model.img_model.img_flipud:
+            return 'Mask (*.mask);;Vertically flipped mask (*.npy *.edf)'
+        return 'Mask (*.mask);;Mask (*.npy *.edf)'
```
I am not sure I understand this logic, in the current implementation img_flipud is always false. I honestly would not try to track this, and always have the option to save normal mask or flipped masked.

t20100 replied:
`img_flipud` is `True` for images loaded with PIL, fabio and lambda loaders and `False` for the others (spe, karabo, hdf5), so I handled this automatically.

But indeed, it's probably not worth it and can be more confusing than explicit options. I'll update that.

t20100 commented:
> _load_mask can be outside of the mask class (not dependent on class) and be renamed to read_mask_file

In which module should it go then? `dioptas.model.util.HelperModule`?

I didn't lint this PR. Which linter do you use?

CPrescher commented:
I am also unsure where to put it, but looking into the class, there are already static methods already on the MaskModel class, maybe just make it static and leave it inside, otherwise probably also OK inside of the MaskModel module

I used default auto-formatting in PyCharm. However, looking into it, I found that for:

def function(default_param="blub"):

no space is used, but for:

def function(default_typed_param: bool = True):

the auto-formatting wants to use spaces. I guess it is a matter of preference:), both are fine for me

t20100 commented on dioptas/controller/MaskController.py:
```diff
@@ -370,35 +375,46 @@ def cosmic_btn_click(self):
-                                    os.path.join(self.model.working_directories['mask'],
-                                                 img_filename + '.mask'),
-                                    filter='Mask (*.mask)')
+        filename, selected_filter = QtWidgets.QFileDialog.getSaveFileName(
```
I replaced `save_file_dialog` with a direct call to `QtWidgets.QFileDialog.getSaveFileName` to retrieve the selected filter. And I use this to toggle the vertical flip. This paves the way to support the same file extensions for both options.

From what I tested with PyQt5 and PyQt6, a 2-tuple is always returned and the call to `set_extension` is not needed

https://github.com/Dioptas/Dioptas/blob/772f6db87a7a4a929de7c90ca58c1c73f81ef6c1/dioptas/widgets/UtilityWidgets.py#L199-L205

t20100 commented on dioptas/controller/MaskController.py:
```diff
@@ -370,35 +375,46 @@ def cosmic_btn_click(self):
     def load_mask_btn_click(self):
-        filename = open_file_dialog(self.widget, caption="Load mask data",
-                                    directory=self.model.working_directories['mask'], filter='*.mask')
+        filename, selected_filter = QtWidgets.QFileDialog.getOpenFileName(
```
Same here, `open_file_dialog` wrapper no longer seems to be needed.

t20100 commented:
I updated the PR to address the comments:

- Removed `img_flipud` logic 
- commit b9e71a00c618a1f4382dd02c3a217612e7b6fcc4 : Use 2 explicit options in the dialogs instead: `Mask` and `Vertically flipped mask`
- commit 406e49316c450ece556022a4679369dc9cc3eb0d :
  - Renamed `_load_mask` to `read_mask_file` and made static
  - documented `flipud` parameter
  - Set default to `read_mask_file`'s `flipud` parameter
  - Fixed typed argument linting: white space around `=` looks to be the standard.
- Updated tests to accommodate with the changes

t20100 commented on dioptas/controller/MaskController.py:
```diff
@@ -33,6 +33,11 @@
 
 class MaskController(object):
+
+    DEFAULT_MASK_FILTER = 'Mask (*.mask)'
```
I hesitate to add `.npy` and `.edf` also to the default:
```suggestion
    DEFAULT_MASK_FILTER = 'Mask (*.mask *.npy *.edf)'
```
The remaining of the code should work fine.

CPrescher commented:
Thanks, i think it looks good. I think the current separation of defaulting to only see *.mask files upon loading is ok. For saving I am not sure whether *.npy and *.edf should be combined. This would currently mean you manually have to delete the npy ending and type in edf as a file ending, which is odd and an unusual user experience - for loading they could still be combined row though.

t20100 commented:
I updated the dialog options as you proposed:
- one filter per format for saving (indeed much better, thanks!)
- multiple format per option for loading.

Does something need to be changed for the batch?

t20100 commented on dioptas/controller/MaskController.py:
```diff
@@ -370,35 +374,56 @@ def cosmic_btn_click(self):
+            caption="Save mask data",
+            directory=os.path.join(self.model.working_directories['mask'],
+                                   img_filename + '.mask'),
+            filter=';;'.join([
+                self.DEFAULT_MASK_FILTER,
```
For completeness, shouldn't we add?

```suggestion
            filter=';;'.join([
                self.DEFAULT_MASK_FILTER,
                "Mask (*.npy)",
                "Mask (*.edf)",
```

t20100 commented on dioptas/controller/MaskController.py:
```diff
@@ -370,35 +374,56 @@ def cosmic_btn_click(self):
+            caption="Load mask data",
+            directory=self.model.working_directories['mask'],
+            filter=';;'.join([
+                self.DEFAULT_MASK_FILTER,
```
Same here:
```suggestion
                self.DEFAULT_MASK_FILTER,
                "Mask (*.npy *.edf)",
```

t20100 commented on dioptas/controller/MaskController.py:
```diff
@@ -370,35 +374,56 @@ def cosmic_btn_click(self):
+            caption="Add mask data",
+            directory=self.model.working_directories['mask'],
+            filter=';;'.join([
+                self.DEFAULT_MASK_FILTER,
```
```suggestion
                self.DEFAULT_MASK_FILTER,
                "Mask (*.npy *.edf)",
```

CPrescher commented:
Thanks. If you do not have anything more, you want to add, I would merge the pull request. Is this ok with you?

t20100 commented:
It's fine for me as it is.
Thanks for the review!
